hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -241,39 +240,49 @@ object HoodieSparkSqlWriter {
             var internalSchemaOpt = getLatestTableInternalSchema(fs, basePath, sparkContext)
-            if (reconcileSchema && parameters.getOrDefault(DataSourceReadOptions.SCHEMA_EVOLUTION_ENABLED.key(), "false").toBoolean
-              && internalSchemaOpt.isEmpty) {
-              // force apply full schema evolution.
-              internalSchemaOpt = Some(AvroInternalSchemaConverter.convert(schema))
-            }
-            if (reconcileSchema) {
-              schema = lastestSchema
-            }
-            if (internalSchemaOpt.isDefined) {
-              // Apply schema evolution.
-              val mergedSparkSchema = if (!reconcileSchema) {
-                AvroConversionUtils.convertAvroSchemaToStructType(AvroSchemaEvolutionUtils.canonicalizeColumnNullability(schema, lastestSchema))
+
+            val writerSchema: Schema =
+              if (reconcileSchema) {
+                // In case we need to reconcile the schema and schema evolution is enabled,
+                // we will force-apply schema evolution to the writer's schema
+                if (enabledSchemaEvolution && internalSchemaOpt.isEmpty) {
+                  internalSchemaOpt = Some(AvroInternalSchemaConverter.convert(sourceSchema))
+                }
+
+                if (internalSchemaOpt.isDefined) {
+                  // Apply schema evolution, by auto-merging write schema and read schema
+                  val mergedInternalSchema = AvroSchemaEvolutionUtils.reconcileSchema(sourceSchema, internalSchemaOpt.get)
+                  AvroInternalSchemaConverter.convert(mergedInternalSchema, latestTableSchema.getName)
+                } else if (TableSchemaResolver.isSchemaCompatible(sourceSchema, latestTableSchema)) {
+                  // In case schema reconciliation is enabled and source and latest table schemas
+                  // are compatible (as defined by [[TableSchemaResolver#isSchemaCompatible]], then we will
+                  // pick latest table's schema as the writer's schema
+                  latestTableSchema
+                } else {
+                  // Otherwise fallback to original source's schema
+                  sourceSchema
+                }
               } else {
-                // Auto merge write schema and read schema.
-                val mergedInternalSchema = AvroSchemaEvolutionUtils.reconcileSchema(schema, internalSchemaOpt.get)
-                AvroConversionUtils.convertAvroSchemaToStructType(AvroInternalSchemaConverter.convert(mergedInternalSchema, lastestSchema.getName))
+                // In case reconciliation is disabled, we still have to do nullability attributes
+                // (minor) reconciliation, making sure schema of the incoming batch is in-line with
+                // the data already committed in the table
+                AvroSchemaEvolutionUtils.canonicalizeColumnNullability(sourceSchema, latestTableSchema)
```

Comment:
Notice:
The avro Schema created by AvroSchemaEvolutionUtils.canonicalizecolumnnullability  has a different namespce with the schema created by AvroConversionUtils.convertStructTypeToAvroSchema.
Avro 1.8.2 is Namespace sensitive.
In the original logic, we will eventually call AvroConversionUtils.convertStructTypeToAvroSchema to maintain namespace consistency

Reply:
@alexeykudinkin : prior to this patch, when `reconcileSchema` is disabled and when no schema evolution is enabled, where do we do canonicalizecolumnnullability?  would you mind pointing me to the code path.

Reply:
That's a good call out @xiarixiaoyao. Do you think it still might be an issue even if all our tests are green (for Spark 2.4)

Reply:
What i meant is that all of our Spark writing is proxied t/h this code and therefore would there be issues due to namespacing in Avro 1.8 (Spark 2.4) we would learn it pretty fast

Reply:
ok that sounds fair. i'm more curious on what different namespaces we might get? by right Hudi should dominate the namespace with some constant like `hoodie_namespace` for e.g., then we should UT cover this.

Reply:
We might get auto-gen'd namespace when we convert incoming `StructType` into `Schema`. Frankly, it's of little variance given how many times we do back-n-forth conversions. We just need to make sure it's not breaking anything

Reply:
@alexeykudinkin  
I have encounter this problem before，So in the original code logic, I finally added avroconversionutils Convertstructtypetoavroschema to maintain the consistency of the namespace. let me check it again with new branch code.
another things:
line 281, we register a writerSchema to kryo, let named this writerSchema as schema1;
line 175 in HoodieSparkUtils we use another schema to convert internal row to avro record,  let named this schema as schema2
schema1 and schema2 has different namespace,  I'm not sure if this will affect performance